### ISSUE TITLE
Pickup and delivery

### ISSUE THREAD

**Issue by N-Wouda** [PyVRP/PyVRP#331]
#### Describe the solution you'd like
That we add support for (simultaneous) pickup and delivery VRPs.

#### Describe alternatives you have considered
None.

#### Additional context
It might be possible to add a new `Parcel` class, which has `origin`, `destination`, and `amount` fields. Currently we only support `origin = 0`, but that could be relaxed.

**Comment by N-Wouda** [comment]
Came up in #5.

**Comment by N-Wouda** [comment]
I have a concatenation scheme for this, using the `DynamicBitset` that I introduced earlier. It should not be too hard to build this in on top of that.

**Comment by N-Wouda** [comment]
There's a bunch of (large) instances available here: https://github.com/cssartori/pdptw-instances.

**Comment by N-Wouda** [comment]
For the next release cycle (so, 0.8.0 or so), I would like to try to get some form of mixed backhaul into PyVRP. Right now we only support deliveries from the depots to the clients, and I would like to also add return shipments from the clients to the depots.

[João Santos et al. (2023)](https://doi.org/10.1016/j.omega.2023.102864) provide an overview of many VRPs with pickup and delivery, including various forms of backhaul.

**Comment by N-Wouda** [comment]
> For the next release cycle (so, 0.8.0 or so), I would like to try to get some form of mixed backhaul into PyVRP. Right now we only support deliveries from the depots to the clients, and I would like to also add return shipments from the clients to the depots.
> 
> [João Santos et al. (2023)](https://doi.org/10.1016/j.omega.2023.102864) provide an overview of many VRPs with pickup and delivery, including various forms of backhaul.

I'm splitting this off to a separate issue, so we can track it for v0.8.0.

**Comment by N-Wouda** [comment]
Mentioned in #486. Is there a lot of demand for this, or is it mostly of interest academically?

**Comment by N-Wouda** [comment]
One-to-one pickup and delivery also gets us clustered VRP (see #470 for some details)

**Comment by ldefour** [comment]
Thanks for the great work!

Would love the see this feature implemented.
OR-Tools has an implementation [here](https://developers.google.com/optimization/routing/pickup_delivery) but unfortunately it can't solve the problem once we have >10 locations and time-window constraints.

**Comment by taimaruuu** [comment]
is there any update on the precedence constraints?

**Comment by leonlan** [comment]
> is there any update on the precedence constraints?

There are no updates on this. We are currently working on other features such as #366.

**Comment by Olivier418** [comment]
> Mentioned in [#486](https://github.com/PyVRP/PyVRP/issues/486). Is there a lot of demand for this, or is it mostly of interest academically?

Just wanted to note that this feature is definitely still in demand! Really impressed with the solver, precedence constraints is the one thing missing for my use case. Any updates are welcome :) Keep up the great work!

**Comment by N-Wouda** [comment]
> Any updates are welcome :)

Great to hear that there's interest @Olivier418! The status of this feature hasn't changed since the last update by Leon at the start of this year - while we are interested in picking this up _eventually_, we currently have no real use cases nor funding to develop pickup and delivery support in PyVRP.

**Comment by Olivier418** [comment]
For anyone stumbling onto this thread in hopes of adding precedence constraints, I found a workaround to get the solver to enforce them. It's not very efficient, but is helpful for a small number of precedence constraints. If this is not the place to post this, let me know :)

To enforce a precedence constraint, you want two things; make sure client A is visited before client B, and force the clients in the same route.

To force the order of A and B, you can add a new load dimension per precedence constraint. All vehicles have a capacity of 1 on this load dimension, and all vertices except A and B contribute 0 to this dimension. A has a delivery of 1, and B has a pickup of 1. This way, A must be visited before B, because the capacity of 1 will be exceeded otherwise:

![Image](https://github.com/user-attachments/assets/4a31dbb6-6657-4798-891a-7fbc7639acce)

To force A and B in the same route, we add n dummy clients next to the depot (where n is the total number of vehicles available). These clients have distance and duration 0 from the depot, and must be visited before any of the acutal clients. n-1 of these clients have a pickup of 1, and the last vertex has no pickup/delivery. The vehicle passing through this last vertex will be visiting both A and B, since no other vehicle can fit them. 

This can actually be simplified a bit if we only have one vehicle, since the dummy clients will not be necessary in that case. Below is an illustration for 5 vehicles, and 6 precedence constraints:

![Image](https://github.com/user-attachments/assets/a504a02b-07d3-471a-a2e7-aa73ab96acc0)

The red dot is the depot, the black crosses are dummy clients for vehicles doing the precedent constraintd and squares are dummy clients for vehicles not doing a precedence constraint. black edges have duration/distance zero, blue edges are edges in the original graph, and the red edges were orignally directed from the depot to the graph.

This is how it looks like in the pyVRP result plot. You can see the dummy vertices added, and how the solution respects the precedence constraints.

![Image](https://github.com/user-attachments/assets/d5aa69ad-b871-4dd7-b4ae-f6ad4daa8d25)

From my testing, adding these constraints does take a toll on the solver's efficiency. After all, we are adding n*P clients and P load dimensions, where P is the number of precedence constraints. Already for 200 clients and 12 precedence-constraints, i noticed an almost three-fold increase in time per iteration. The solver also took a very long time to even find a solution. Maybe this can be improved by providing the solver with an initial solution so it properly navigates through the dummy clients, but this feature isn't availble yet either.

If you only have a few precedence constraints or a single vehicle, this trick might be helpful.

**Comment by N-Wouda** [comment]
Hi @Olivier418, that looks amazing! Do you know which of the two - the extra dummy clients or load dimensions - makes it particularly difficult for PyVRP to find an initial feasible solution?

**Comment by Olivier418** [comment]
Hey @N-Wouda, I'm not quite sure. It should be quite easy to investigate, since the dummy clients can be left out for a TSP. I'll work on some code to benchmark it, but I'm quite busy so it might take a while. I'll get back to you!

**Comment by Olivier418** [comment]
I managed to get around to it earlier than I expected :) Here's the results:

Varying the number of vehicles and precedence constraints (which I will call pickup-deliveries from now on), we can create a figure that shows the solver's convergence either with or without precedence constraints:

![Image](https://github.com/user-attachments/assets/c3265d0f-e2e4-4c8c-8b0d-ef22f7446901)

From this, it's definitely noticeable that the solver is slowed down a lot by an increasing number of vehicles/pickup-deliveries, and that it takes longer to find an initial feasible solution. 

![Image](https://github.com/user-attachments/assets/a5b16e6e-2b1c-4688-b89e-28b5a13a84de)

In the above figure, we compare the increase in average time per iteration step when adding pickup-deliveries across the number of vehicles, number of pickup deliveries (which equals the number of additional load dimensions) and the number dummy clients. The number of vehicles seems to have no effect at all, and the number of pickup-deliveries / dimensions seems to have the greatest effect on the solvers speed (with an increase of almost 4x in iteration time). The number of dummy clients shows a little correlation too, but I think this is a statistical artifact caused by the fact that the number of dimensions and dummy clients are linearly dependent. 

![Image](https://github.com/user-attachments/assets/9ac9fcf2-2d2c-4061-9948-737902737169)

This is the actual info you asked for ;) This shows the increase in time/nr of iterations until an initial feasible solution is found across the same parameters as before. I think this data is a bit more difficult to interpret, but it seems to me like the number of dummy-clients is the main culprit for this issue. The others show some correlation, but not as much. Would love to hear your interpretation of this though!

**Comment by N-Wouda** [comment]
@Olivier418 thanks for the detailed study! I'm trying to interpret your results, but I'm somewhat unsure about the distinction between the shaded and solid dots in your second and third figures. What do those mean?

**Comment by Olivier418** [comment]
@N-Wouda The low-opacity dots are individual test runs on instances of the same number of vehicles and pickup-deliveries, while the fully opaque (solid) dots show the average across those runs. If you're mainly interested in the overall trend and want to filter out noise or outliers, you can focus on the solid dots.

**Comment by N-Wouda** [comment]
@Olivier418 OK, looking at your plots, I have a few (not very deep) observations. The first is that both the additional load dimensions and the dummy vertices needed to model PDs seem to increase the duration of a single iteration, which almost surely is caused by a longer local search procedure. Since the two are linearly dependent, as you say, due to the PD. Perhaps obviously, increasing the number of PDs results in longer time until feasible. I think that makes sense, too: modelling PDs this way is somewhat of a workaround the solver doesn't really understand, and doing more of this isn't any easier.

Then, in the first figure, the TSP situation is much easier because the decision to put PDs on the same route is already made. The VRP cases all seem roughly equally hard, which supports the idea that most of the difficulty is in the local search: any PD-improving moves have to be based on node operators, which work on client pairs, and are thus independent of the _number_ of routes/vehicles. You already noticed this in your second and third figures, which show no relation to the number of vehicles in the instance. That is expected.

Thanks for investigating further!

**Comment by N-Wouda** [comment]
Are your vehicles all the same @Olivier418? In that case you might be able to select one up front to visit A and B, by giving all the others an `initial_load` equal to one in the appropriate load dimension. This'd replace the dummy vertices.

**Comment by Olivier418** [comment]
@N-Wouda The vehicles can be different in my example, unfortunately. But I like your idea, I think that should work for different vehicles for sure. Thanks for the suggestion!

I actually thought of another way to structure the grid of dummy clients by using optional clients, which significantly reduces the number of edges and removes the symmetry. I won't have time to work on that for a while, but just to say, there are probably better ways to do this.

**Comment by N-Wouda** [comment]
We'll need to think a bit about how to generalise the `LoadSegment` to handle pickup and delivery between clients (rather than just clients and depots). That might not be too hard, except for `LoadSegment::finalise`. How do we want to handle reloading and pickup and delivery?

**Comment by N-Wouda** [comment]
Something like (inspired by VROOM):
```cpp
struct Shipment
{
  struct ShipmentStep
  {
    size_t location;
    Duration twEarly;
    Duration twLate;
    Duration serviceDuration;
  };

  ShipmentStep pickup;
  ShipmentStep delivery;
  std::vector<Load> amount;
  Cost prize;
  bool required;
};
```

Constructor interface something like:
```cpp
Shipment(
  size_t pickupLocation,
  Duration pickupEarly,
  Duration pickupLate,
  Duration pickupService,
  size_t deliveryLocation,
  etc...
);
```
additionally, probably a constructor that takes the `ShipmentStep`s as well.

(I don't like `ShipmentStep`, and need to think about a better name for it)

**Comment by N-Wouda** [comment]
From [Hiermann and Schiffer (2026)](https://doi.org/10.1016/j.ejor.2026.01.037):

<img width="1040" height="709" alt="Image" src="https://github.com/user-attachments/assets/a6af19ad-7f47-4eac-ab43-cfc2c205e57f" />

Need to think about the interaction of this with reloading/finalising client demand, but that should be OK. We'll probably want different fields for these anyway, and only have them interact for (excess) load.

**Comment by N-Wouda** [comment]
Possibly interesting for evaluation: https://github.com/siemterbraake/vrppdl-algorithms-instances/tree/main. Don't know if we can do this already (scheduled linehauls look challenging), but might be worth spending a minute to think about.

(and paper [here](https://www.sciencedirect.com/science/article/pii/S0377221725005958))

**Comment by leonlan** [comment]
The 2024 OGC challenge has a bunch of heterogeneous PDP instances. See https://github.com/ogchallenge/ogc and https://ogchallenge.github.io/ogc/#2024.

> A heterogeneous-fleet pickup-and-delivery problem with time windows/deadlines, capacity constraints, and limited rider availability. Riders can be WALK / BIKE / CAR, with different speed, capacity, service time, fixed cost, and variable distance cost. Each bundle chooses a rider type, a set of orders, pickup order, and delivery order; all orders must be served, each rider gets at most one bundle, and the objective is average delivery cost per order.
> 
> A distinctive constraint: for a bundle, all pickups must be completed before any deliveries. So it is a structured pickup-and-delivery / meal-delivery bundling problem, not a general pickup-delivery route where pickups and dropoffs can freely interleave

**Comment by leonlan** [comment]
Can benchmark against SISR on PDPTW here: https://github.com/hankarudova/open-source-sisr-routing/tree/pdptw

Paper: https://link.springer.com/chapter/10.1007/978-3-032-27242-3_22

---

https://github.com/Yamp/pdp_lns

https://github.com/cssartori/pdptw-instances/tree/master/solutions

**Comment by N-Wouda** [comment]
PD moves:
- Update the Exchange template to allow moves involving shipments, but only if the shipment precedence constraints are not violated.
- Same for SwapTails; fine as long as no precedence constraint is violated;
- Add `RelocateShipment` move that relocates a shipment pair to somewhere else. I need to work out the details further. Perhaps we can base this on `RelocatePair` of [Pacheco et al. (2022)](https://doi.org/10.1287/trsc.2022.1176).
- Add a `SwapShipment` move that exchanges two shipments.

**Comment by N-Wouda** [comment]
General search idea:

In `SearchSpace` we now have `clientOrder()`, which is the order in which the LS goes through the clients. We also have promising client nodes. I think we can expand that to activities of clients and shipments. In practice that'll be clients and pickups, since we do not need to duplicate shipments here - the move for pickups is the same as for deliveries.

The granular neighbourhood should have both entries, since a client or pickup can be close to some shipment's delivery node, but not necessarily also its corresponding pickup. I think we can do this as either a matrix or map of activities (client, pickup) to lists of activities (clients, pickups, deliveries).

In the search we then have that a node $U$ is always either a client or a pickup. The node $V$ can be anything. Since $U$ is never a delivery, we do not need to handle the case where $U$ is a delivery in the operators.

**Comment by N-Wouda** [comment]
This feature is now functional in #1098 for anyone here that'd like to try it. We're still working on better performance, documentation, and ironing out the last kinks. Should be released next month as part of 0.14 (https://github.com/PyVRP/PyVRP/issues/945).

**Comment by N-Wouda** [comment]
Just merged into main as https://github.com/PyVRP/PyVRP/commit/856a144b44138a27dc49c80582f29490173b7b17. Will be released next week as part of 0.14.
